sympy/matrices/random.py

```diff
@@ -0,0 +1,2378 @@
+from ..core import I as _i
+from ..core.mul import Mul as _multiply
+from ..core.random import sample as _smple
+from ..core.sympify import _sympify
+from ..functions import sqrt as _sqrt, re as _re, im as _im, \
+    transpose as _t, adjoint as _c
+from ..tensor import shape as _shape
+from .dense import eye as _eye
+from .inverse import _inv
+
+__all__ = 'projection', 'jordan', 'transposition', \
+          'permutation', 'elementary', 'rotation', 'reflection', \
+          'diagonal_normal', 'jordan_normal', 'isometry_normal', \
+          'triangular', 'invertible', 'singular', \
+          'idempotent', 'nilpotent', 'diagonalizable', 'trigonalizable', \
+          'orthogonal', 'unitary', 'normal', \
+          'symmetric', 'hermite', 'square', \
+          'regular_to_singular', 'complex_to_real'
+
+
+_EPS = 1e-15
+
+
+# === random number generator functions ===
+
+
+def _sample(scalars, k=None):
+    _k = k or 1
+    if hasattr(scalars, 'sample'):
+        smpl = scalars.sample(_k)
+    else:
+        smpl = _smple(scalars, _k)
+    return smpl if k else smpl[0]
+
+
+# === helper ===
+
+
+def _cs(scalar, real=True):
+    if isinstance(scalar, (tuple, list)) and len(scalar) == 2:
+        c, s = scalar
+    elif _is_complex(scalar):
+        c, s = _re(scalar), _im(scalar)
+    else:
+        c, s = scalar, _sample((-1, 1)) * _sqrt(1 - scalar ** 2)
```

Comment:
This function name is quite cryptic. It could do with at least comments or a docstring but probably a more descriptive name e.g. one that describes what it returns or does.

The signature is quite confusing since it accepts either a tuple or a "scalar" and then returns either a tuple or a complex number. The function is called in only two places only one of which uses the `real` argument which makes me wonder if it really makes sense for this combination of functionality to be grouped together in a single function.

Reply:
You're right the functions signature is a mess. I reworked depending functions like `isometry_normal` and `unitary` to rely only on `rotation`. So I was able to moved the remaining parts of `_cs` into `rotation`. Finally, I was able to remove the `real` argument.

Hope this makes the concept of the whole module more clear. 
As it constructs _high level matrices_ (like `orthogonal`, `unitary` or `normal`) by _conjugating_ a _normal form_ (here `isometry_normal`) with a product of _basis matrices_ (of type `rotation`).

And now the default behaviour of `scalars` arguments are set in `rotation`.